app/db/interface.py

```diff
@@ -139,12 +140,85 @@ def _remove_soft_deletes(self, payments):
+        """
+        try:
+            payments = self._query_payments(payment_attributes={"user_id":user_id})
+            if 'payment_id' in payment_attributes: #for set-default action
+                payment_id_to_be_updated = payment_attributes['payment_id']
+                payment_to_be_updated = None
+                for payment in payments:
+                    if payment.id == payment_id_to_be_updated: #no need to check if is_removed = False because the query filters it
+                        payment_to_be_updated = payment
+                if payment_to_be_updated != None:
+                    payment_to_be_updated.is_default = True
+                    #making sure other payments have default as false
+                    other_payments = [payment for payment in payments if payment != payment_to_be_updated]
+                    for other_payment in other_payments:
+                        other_payment.is_default = False
+                    #Committing all changes
+                    self.db.session.commit()
+                    return True
+                else:
+                    return False
+            elif 'amount' in payment_attributes: #for charge action
+                default_payment = None
+                for payment in payments:
+                    if payment.is_default == True:
+                        default_payment = payment
+                if default_payment != None:
+                    update_payment_flag = True
+
+                    if default_payment.payment_type == 'paypal' and not default_payment.details.is_linked:
+                        error_msg = 'Invalid request: Default Payment for this user_id: '+ str(user_id)+' (Paypal) is not linked.'
+                        raise DataValidationError(error_msg)
+                    elif default_payment.payment_type != 'paypal' and self.is_expired(default_payment.details.expires):
+                        error_msg = 'Invalid request: Default Payment for this user_id: '+ str(user_id)+' ('+default_payment.payment_type+') is expired'
+                        raise DataValidationError(error_msg)
+
+                    default_payment.charge_history = default_payment.charge_history + payment_attributes['amount']
+                    self.db.session.commit()
+                    return True
+                else:
+                    error_msg = 'Invalid request: Default Payment for this user_id: '+ str(user_id)+' not found. Please update the default_payment first.'
+                    raise DataValidationError(error_msg) #PaymentNotFoundError
+            else:
+                raise DataValidationError('Invalid request: The request body contains bad or missing data.')
+        except PaymentServiceQueryError as e:
+            error_msg = 'Payments not found for the user_id: '+ str(user_id)
+            raise DataValidationError(error_msg) #PaymentNotFoundError
+
+    def is_expired(self, exp_date):
+        month = int(exp_date[:2]) + 1
```

Comment:
Might be a good idea to wrap any work with indices in a `try/except` block so that we can catch a poorly formatted date and avoid having the backend throw an HTTP 500 error.